chains/btc/listener/event-handlers.go

```diff
@@ -0,0 +1,142 @@
+// The Licensed Work is (c) 2022 Sygma
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package listener
+
+import (
+	"math/big"
+	"strconv"
+
+	"github.com/btcsuite/btcd/btcjson"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/sygmaprotocol/sygma-core/relayer/message"
+)
+
+type Deposit struct {
+	// ResourceID used to find address of handler to be used for deposit
+	ResourceID [32]byte
+	// Address of sender (msg.sender: user)
+	SenderAddress string
+	// Additional data to be passed to specified handler
+	Amount *big.Int
+	Data   string
+}
+
+type DepositHandler interface {
+	HandleDeposit(
+		sourceID uint8,
+		depositNonce uint64,
+		resourceID [32]byte,
+		amount *big.Int,
+		data string,
+		blockNumber *big.Int,
+	) (*message.Message, error)
+	//FetchEvents(startBlock *big.Int) ([]btcjson.TxRawResult, error)
+}
+
+type FungibleTransferEventHandler struct {
+	depositHandler DepositHandler
+	domainID       uint8
+	log            zerolog.Logger
+	conn           Connection
+	msgChan        chan []*message.Message
+	bridge         string
+}
+
+func NewFungibleTransferEventHandler(logC zerolog.Context, domainID uint8, depositHandler DepositHandler, msgChan chan []*message.Message, conn Connection, bridge string) *FungibleTransferEventHandler {
+	return &FungibleTransferEventHandler{
+		depositHandler: depositHandler,
+		domainID:       domainID,
+		log:            logC.Logger(),
+		conn:           conn,
+		msgChan:        msgChan,
+		bridge:         bridge,
+	}
+}
+
+func (eh *FungibleTransferEventHandler) HandleEvents(blockNumber *big.Int) error {
+	domainDeposits := make(map[uint8][]*message.Message)
+	evts, err := eh.FetchEvents(blockNumber)
+	if err != nil {
+		eh.log.Error().Err(err).Msg("Error fetching events")
+		return err
+	}
+	for evtNumber, evt := range evts {
+		func(evt btcjson.TxRawResult) {
+			defer func() {
+				if r := recover(); r != nil {
+					log.Error().Msgf("panic occured while handling deposit %+v", evt)
+				}
+			}()
+
+			d, isDeposit, err := DecodeDepositEvent(evt, eh.conn, eh.bridge)
+			if err != nil {
+				log.Error().Err(err).Msgf("%v", err)
+				return
```

Comment:
Would be nice to return these errors.
If you make the func(evt btcjson.TxRawResult) be func(evt btcjson.TxRawResult) error
Then you can just return from this function